confidant/services/jwkmanager.py

```diff
@@ -0,0 +1,97 @@
+import jwt
+import json
+
+from jwcrypto import jwk
+from typing import Any, Dict, Optional
+from hashlib import sha1
+from cryptography.x509 import load_pem_x509_certificate
+from cryptography.hazmat.primitives import serialization
+from confidant.settings import CERTIFICATE_AUTHORITIES
+
+
+class JWKManager:
+    def __init__(self) -> None:
+        self._keys = jwk.JWKSet()
+        self._public_keys = {}
+        self._token_cache = {}
+        self._payload_cache = {}
+
+        self._load_certificate_authorities()
+
+    def _load_certificate_authorities(self) -> None:
+        if CERTIFICATE_AUTHORITIES:
+            for ca in CERTIFICATE_AUTHORITIES:
+                self.set_key(ca['name'], ca['key'],
+                             passphrase=ca['passphrase'])
+
+    def set_key(self, kid: str, private_key: str,
+                passphrase: Optional[str] = None,
+                encoding: str = 'utf-8') -> str:
+        if passphrase:
+            passphrase = passphrase.encode(encoding)
+        key = jwk.JWK()
+        key.import_from_pem(private_key.encode(encoding),
+                            password=passphrase, kid=kid)
+        if not self._keys.get_key(kid):
+            self._keys.add(key)
+        return kid
+
+    def get_jwt(self, kid: str, payload: dict) -> str:
+        key = self._keys.get_key(kid)
+        if not key:
+            raise ValueError('This private key is not stored!')
+
+        payload_hash = sha1(
+            json.dumps(payload, sort_keys=True).encode('utf-8')
+        ).hexdigest()
+        if kid not in self._token_cache:
+            self._token_cache[kid] = {}
+
+        if payload_hash not in self._token_cache[kid]:
+            self._token_cache[kid][payload_hash] = jwt.encode(
+                payload=payload,
+                headers={'kid': kid},
+                key=key.export_to_pem(private_key=True, password=None),
```

Comment:
Shouldn't we allow customized password here during export. because on line 30-31, we check whether the passphrase is used when load from pem.

or we might remove passphrase check in 30-31 if we know we do not use password to wrap the key when export?

Reply:
On line 30-31 we use the passphrase to import into the JWKSet, that is where we're decrypting the key if provided as encrypted. On line 54 we're exporting it unencrypted so we can send it as input to the `jwt.encode()` function, I don't think there's an option to call it with an encrypted key, so we need to export it without a passphrase.

Also on 54 we're explicitly using `password=None` since the export function requires it to be done that way when you want no password when exporting.

Reply:
Got it. So you mean when we load from pem, it is possble that we need a password to decode, but when we export, we never encrypt the private key bytes.

the reason I am asking is in our test it looks like we are not testing set_key with a password, not sure whether it is a possible scenario tho

Reply:
Yes that's correct. We are not testing right now but it's a good point, I can add another scenario using a private key with password.